esx_service/vmdk_ops.py

```diff
@@ -764,24 +773,45 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     Returns None (if all OK) or error string
     """
     logging.debug("config_path=%s", config_path)
-    vm_datastore = get_datastore_name(config_path)
+    vm_datastore_url = get_datastore_url_from_config_path(config_path)
+    vm_datastore = vmdk_utils.get_datastore_name(vm_datastore_url)
+    if not datastore_path_exist(vm_datastore):
+        # path /vmfs/volumes/vm_datastore_name does not exist
+        # the possible reason is vm_datastore_name which got from
+        # datastore cache is invalid(old name)
+        # need to refresh cache
+        vmdk_utils.init_datastoreCache()
+        vm_datastore = vmdk_utils.get_datastore_name(vm_datastore_url)
+
     error_info, tenant_uuid, tenant_name = auth.get_tenant(vm_uuid)
     if error_info:
         return err(error_info)
     if tenant_name == auth.DEFAULT_TENANT:
         # for DEFAULT tenant, set default_datastore to vm_datastore
+        default_datastore_url = vm_datastore_url
         default_datastore = vm_datastore
     else:
         # if default_datastore is not set for tenant,
-        # default_datastore will be set to None
-        error_info, default_datastore = auth_api.get_default_datastore(tenant_name)
-        # if get default_datastore fails or default_datastore is not specified,
-        # use vm_datastore
-        if error_info or not default_datastore:
+        # default_datastore_url will be set to None
+        error_info, default_datastore_url = auth_api.get_default_datastore(tenant_name)
+        # if get_default_datastore fails or default_datastore_url is not specified,
+        # use vm_datastore_url
+        if error_info or not default_datastore_url:
+            default_datastore_url = vm_datastore_url
             default_datastore = vm_datastore
+        else:
+             default_datastore = vmdk_utils.get_datastore_name(default_datastore_url)
+             if not datastore_path_exist(default_datastore):
```

Comment:
Totally agree - cache manipulation (refresh on demand) is the API's responsibility, not the API user's.